File: app/src/main/java/com/example/client/screens/home/fragment/HomeFragment.java
package com.example.client.screens.home.fragment;

import android.content.Intent;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.RelativeLayout;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;
import androidx.viewpager2.widget.ViewPager2;

import com.bumptech.glide.Glide;
import com.example.client.R;
import com.example.client.app.Constants;
import com.example.client.models.banner.BannerModel;
import com.example.client.models.branch.BranchModel;
import com.example.client.models.category.CategoryModel;
import com.example.client.models.event.Event;
import com.example.client.models.profile.ProfileModel;
import com.example.client.screens.branch.BranchActivity;
import com.example.client.screens.home.item.BannerItem;
import com.example.client.screens.home.item.HomeCategoryItem;
import com.example.client.screens.home.present.HomePresent;
import com.example.client.screens.product.activity.ProductActivity;
import com.example.client.screens.profile.manager_info.ManagerInfoActivity;
import com.google.android.material.tabs.TabLayout;
import com.google.android.material.tabs.TabLayoutMediator;

import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;
import org.greenrobot.eventbus.ThreadMode;

import java.text.DateFormat;
import java.text.MessageFormat;
import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.List;
import java.util.Locale;


public class HomeFragment extends Fragment implements View.OnClickListener, IHomeView {
    private RecyclerView recyclerViewIcon;
    private ViewPager2 pagerBanner;
    private SwipeRefreshLayout refreshLayout;

    private RecyclerView recyclerViewHighLight;
    private TextView moreHighLight;
    private TextView titleHighLight;

    private RecyclerView recyclerViewNew;
    private TextView moreNew;
    private TextView titleNew;
    private TextView wellcome;
    private LinearLayout profile;
    private TabLayout tabDots;
    private ImageView avatar;
    private ImageView banner_empty,high_light_empty,new_empty;
    private TextView tvBranchName, tvBranchAddress, tvOrderAddress;
    private RelativeLayout rllChangeBranch;

    private HomePresent hPresent;
    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_home,null);
        recyclerViewIcon = view.findViewById(R.id.recyclerViewIcon);
        pagerBanner = view.findViewById(R.id.pagerBanner);
        refreshLayout = view.findViewById(R.id.container);
        avatar = view.findViewById(R.id.avatar);
        recyclerViewHighLight = view.findViewById(R.id.recyclerViewHightLight);
        moreHighLight = view.findViewById(R.id.moreHightLight);
        titleHighLight = view.findViewById(R.id.titleHightLight);

        recyclerViewNew = view.findViewById(R.id.recyclerViewNew);
        moreNew = view.findViewById(R.id.moreNew);
        titleNew = view.findViewById(R.id.titleNew);
        tabDots = view.findViewById(R.id.tabDots);
        wellcome = view.findViewById(R.id.wellcome);
        profile = view.findViewById(R.id.profile);
        banner_empty = view.findViewById(R.id.banner_empty);
        high_light_empty = view.findViewById(R.id.high_light_empty);
        new_empty = view.findViewById(R.id.new_empty);
        tvBranchName = view.findViewById(R.id.tv_branch_name);
        tvBranchAddress = view.findViewById(R.id.tv_branch_address);
        rllChangeBranch = view.findViewById(R.id.rll_change_branch);
        tvOrderAddress = view.findViewById(R.id.tv_order_address);

        hPresent = new HomePresent(this);

        refreshLayout.setOnRefreshListener(() -> {
            hPresent.getProductsHighLightFromService();
            hPresent.getProductNewFromService();
            hPresent.getBranchFromRes();
            hPresent.getCategoriesFromService();
            hPresent.getListBannerFromService();
            refreshLayout.setRefreshing(false);
        });



        moreHighLight.setOnClickListener(this);
        moreNew.setOnClickListener(this);
        profile.setOnClickListener(this);

        hPresent.getUserFromRes();
        hPresent.getProductsHighLightFromService();
        hPresent.getProductNewFromService();
        hPresent.getBranchFromRes();
        hPresent.getCategoriesFromService();
        hPresent.getListBannerFromService();
        rllChangeBranch.setOnClickListener(this);

        return view;
    }

    @Override
    public void onStart() {
        super.onStart();
        if (!EventBus.getDefault().isRegistered(this))
            EventBus.getDefault().register(this);
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    public void onEvent(Event event) {
        switch (event.getKey()) {
            case Constants.EventKey.CHANGE_BRANCH:
                hPresent.getBranchFromRes();
                break;
            case Constants.EventKey.UPDATE_PROFILE_AVATAR:
            case Constants.EventKey.UPDATE_LOCATION:
            case Constants.EventKey.UPDATE_PROFILE_INFO:
                hPresent.getUserFromRes();
                break;
        }
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        EventBus.getDefault().unregister(this);
    }

    @Override
    public void onClick(View v) {
        switch (v.getId()){
            case R.id.moreHightLight:

                break;
            case R.id.moreNew:

                break;
            case R.id.profile:
                startActivity(new Intent(getActivity(), ManagerInfoActivity.class));
                break;
            case R.id.rll_change_branch:
                startActivity(new Intent(getActivity(), BranchActivity.class));
                break;
        }
    }

    @Override
    public void showCategories(List<CategoryModel> items) {
        GridLayoutManager manager = new GridLayoutManager(getContext(), 2, GridLayoutManager.HORIZONTAL, false);
        recyclerViewIcon.setLayoutManager(manager);
        HomeCategoryItem item = new HomeCategoryItem(getContext(), items, categoryModel -> startActivity(ProductActivity.Companion.newInstance(getContext(),categoryModel, Constants.Method.CATEGORY)));
        recyclerViewIcon.setAdapter(item);
    }

    @Override
    public void showBanners(List<BannerModel> items) {
        if(items == null || items.size() == 0){
            banner_empty.setVisibility(View.VISIBLE);
            pagerBanner.setVisibility(View.GONE);
        }
        else {
            banner_empty.setVisibility(View.GONE);
            pagerBanner.setVisibility(View.VISIBLE);
            BannerItem bannerItem = new BannerItem(getActivity(),items);
            pagerBanner.setAdapter(bannerItem);
            new TabLayoutMediator(tabDots,pagerBanner,(tab, position) -> {

            }).attach();
        }

    }

    @Override
    public void showBranchInfo(BranchModel branch) {
        tvBranchName.setText(branch.getName());
        tvBranchAddress.setText(branch.getAddress());
    }

    @Override
    public void showUserInfo(ProfileModel profile) {
        DateFormat format = new SimpleDateFormat("HH", Locale.getDefault());
        int hour = Integer.parseInt(format.format(Calendar.getInstance().getTime()));
        String greet = "";
        if (hour >= 4 && hour < 11)
            greet = "buổi sáng, ";
        else if (hour >= 11 && hour < 13)
            greet = "buổi trưa, ";
        else if (hour >= 13 && hour < 18)
            greet = "buổi chiều, ";
        else
            greet = "buổi tối, ";
        wellcome.setText(MessageFormat.format("Chào {0}{1}", greet, profile.getFullname()));
        Glide.with(this)
                .asBitmap()
                .placeholder(R.drawable.avatar_default)
                .load(profile.getAvatar())
                .into(avatar);
        tvOrderAddress.setText(profile.getAddress());
    }

    @Override
    public void toBranchScreen() {
        startActivity(new Intent(getActivity(), BranchActivity.class));
    }

}

File: app/src/main/java/com/example/client/screens/home/fragment/HomeFragment.kt
package com.example.client.screens.home.fragment

import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout.OnRefreshListener
import com.bumptech.glide.Glide
import com.example.client.R
import com.example.client.base.BaseFragmentMVP
import com.example.client.dialog.AddToCartDialog
import com.example.client.dialog.OptionAddToCartListener
import com.example.client.models.branch.BranchModel
import com.example.client.models.cart.CartProductModel
import com.example.client.models.category.CategoryModel
import com.example.client.models.order.OrderLocation
import com.example.client.models.category.HomeSectionModel
import com.example.client.models.product.ProductModel
import com.example.client.models.profile.ProfileModel
import com.example.client.screens.branch.BranchActivity
import com.example.client.screens.category.activity.CategoryActivity
import com.example.client.screens.category.item.HomeCategoryItem
import com.example.client.screens.category.parent.SuperCategoryActivity
import com.example.client.screens.home.item.HomeSectionItem
import com.example.client.screens.home.present.HomePresent
import com.example.client.screens.home.present.IHomePresent
import com.example.client.screens.main.activity.MainActivity
import com.example.client.screens.map.activity.MapsActivity
import com.example.client.screens.product.activity.ProductActivity
import com.example.client.screens.profile.manager_info.activity.ManagerProfileActivity
import kotlinx.android.synthetic.main.fragment_home.*
import java.text.DateFormat
import java.text.SimpleDateFormat
import java.util.*

class HomeFragment : BaseFragmentMVP<IHomePresent>(), View.OnClickListener, IHomeView, OnRefreshListener, OptionAddToCartListener {

    companion object {
        @JvmStatic
        fun newInstance() = HomeFragment()
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_home, null)
    }

    override fun bindData() {
        mPresenter?.run {
            binData()
            getCategories()
            getHomeSections()
        }
    }

    override fun bindEvent() {
        swipe_refresh.setOnRefreshListener(this)
        lnl_profile.setOnClickListener(this)
        rll_change_branch.setOnClickListener(this)
        lnl_order_location.setOnClickListener(this)
    }

    override fun onClick(v: View) {
        when (v.id) {
            R.id.lnl_profile -> startActivity(ManagerProfileActivity.newInstance(requireActivity()))
            R.id.rll_change_branch -> startActivity(BranchActivity.newInstance(requireActivity()))
            R.id.lnl_order_location -> startActivity(MapsActivity.newInstance(requireActivity()))
        }
    }

    override fun showCategories(items: List<CategoryModel>) {
        recycler_view_category.visibility = View.VISIBLE
        val manager = GridLayoutManager(context, 4, GridLayoutManager.VERTICAL, false)
        recycler_view_category.layoutManager = manager
        val item = HomeCategoryItem(requireContext(), items, { categoryModel ->
            startActivity(SuperCategoryActivity.newInstance(requireActivity(), categoryModel))
        }, {
            startActivity(CategoryActivity.newInstance(requireActivity()))
        })
        recycler_view_category.adapter = item
    }

    override fun hideCategories() {
        recycler_view_category.visibility = View.GONE
    }

    override fun showBranch(branch: BranchModel) {
        tv_branch_name.text = branch.name
        tv_branch_address.text = branch.address
    }

    override fun showProfile(profile: ProfileModel) {
        profile.let {
            val format: DateFormat = SimpleDateFormat("HH", Locale.getDefault())
            val hour = format.format(Calendar.getInstance().time).toInt()
            val greet = if (hour in 4..10) getString(R.string.greet_morning) else if (hour in 11..12) getString(R.string.greet_lunch) else if (hour in 13..17) getString(R.string.greet_evening) else getString(R.string.greet_night)
            tv_greet.text = getString(R.string.greet_description, greet, it.fullname)
            Glide.with(this)
                    .asBitmap()
                    .placeholder(R.drawable.avatar_default)
                    .load(it.avatar)
                    .into(view_icon)
        }
    }

    override fun showOrderLocation(orderLocation: OrderLocation) {
        tv_order_address.text = orderLocation.address
    }

    override fun toBranchScreen() {
        startActivity(Intent(activity, BranchActivity::class.java))
    }

    override fun showErrorMessage(errMessage: Int) {
        showToastMessage(getString(errMessage))
    }

    override fun showHomeSections(items: List<HomeSectionModel>) {
        recycler_view_section.visibility = View.VISIBLE
        val manager = LinearLayoutManager(context, LinearLayoutManager.VERTICAL, false)
        recycler_view_section.layoutManager = manager
        val item = HomeSectionItem(requireContext(), items, {
                startActivity(ProductActivity.newInstance(requireActivity(), it))
            },
            { prod ->
                startActivity(ProductActivity.newInstance(requireActivity(), prod, true))
            },
            { prod ->
                showAddToCartDialog(prod)
            }
        )
        recycler_view_section.adapter = item
    }

    override fun showLoading() {
        swipe_refresh.isRefreshing = true
    }

    override fun hideLoading() {
        swipe_refresh.run {
            if (isRefreshing) isRefreshing = false
        }
    }

    override val presenter: IHomePresent
        get() = HomePresent(this)

    override fun onBackPress() {
        requireActivity().onBackPressed()
    }

    override fun onRefresh() {
        bindData()
        if (requireActivity() is MainActivity) (requireActivity() as MainActivity).onRefreshOrder()
    }

    private fun showAddToCartDialog(product: ProductModel) {
        val dialog = AddToCartDialog.newInstance(product)
        dialog.setListener(this)
        dialog.show(childFragmentManager)
    }

    override fun addToCart(cartProduct: CartProductModel) {
        mPresenter?.addToCart(cartProduct)
    }

    override fun showProductDetail(product: ProductModel) {
        startActivity(ProductActivity.newInstance(requireActivity(), product, true))
    }
}